File: packages/ios/Tgrok/web/JsRequest.swift
//
//  JsRequest.swift
//  PipiJr
//
//  Created by Yueyu Zhao on 2022/4/23.
//

import Foundation
import SwiftyJSON
import WebKit

class JsRequest: NSObject {
    let webView: WKWebView

    let id: String!
    let clsName: String
    let clsMethod: String
    let params: JSON
    let callback: String
    let origMethod: String

    init(_ webView: WKWebView, _ json: String) {
        self.webView = webView

        let body = JSON(parseJSON: json)
        self.id = body["id"].string
        origMethod = body["method"].stringValue
        let methods = origMethod.components(separatedBy: "@")
        clsName = methods[0]
        clsMethod = methods[1]
        params = body["params"]
        callback = body["callback"].stringValue
    }

    func callback(_ result: String) {
        if nil == self.id {
            return
        }
        var format = "window.drmer && drmer.dequeue('%@', '%@');"
        // result is a JSON string
        if "null" == result || result.starts(with: "{") || result.starts(with: "[") {
            format = "window.drmer && drmer.dequeue('%@', %@);"
        }
        let script = String(format: format, self.id, result)
        DispatchQueue.main.async {
            self.webView.evaluateJavaScript(script, completionHandler: nil)
        }
    }

    func callback(_ result: Bool) {
        self.callback(result ? "true" : "false")
    }

    func callback(_ result: JSON) {
        self.callback(result.description)
    }

    func intParam(_ key: String) -> Int {
        return self.params[key].intValue
    }

    func strParam(_ key: String) -> String {
        return self.params[key].stringValue
    }

    func boolParam(_ key: String) -> Bool {
        return self.params[key].boolValue
    }

    func jsonParam(_ key: String) -> JSON {
        return JSON(parseJSON: self.strParam(key))
    }
}

File: packages/android/app/src/main/java/net/drmer/android/tgrok/web/JsRequest.kt
package net.drmer.android.tgrok.web

import android.os.Build
import com.google.gson.Gson
import org.apache.commons.lang3.StringUtils
import org.json.JSONArray
import org.json.JSONObject

class JsRequest(private val activity: WebViewActivity, body: String) {

    var clsName: String? = null
        private set

    var clsMethod: String = ""

    var id: String? = null
        private set

    var params: JSONObject? = null
        private set

    init {
        val jsonObject = JSONObject(body)
        if (jsonObject.has("method")) {
            val method = jsonObject.optString("method").split("@".toRegex()).toTypedArray()
            clsName = method[0]
            clsMethod = method[1]
            id = jsonObject.optString("id")
            params = jsonObject.optJSONObject("params")
        }
    }

    fun valid(): Boolean {
        return StringUtils.isNotEmpty(clsName) && StringUtils.isNotEmpty(clsMethod)
    }

    fun intParam(key: String): Int {
        return params!!.optInt(key)
    }

    fun strParam(key: String): String {
        return params!!.optString(key)
    }

    fun objParam(key: String): JSONObject {
        return params!!.optJSONObject(key)
    }

    fun arrParam(key: String): JSONArray {
        return params!!.optJSONArray(key)
    }

    fun boolParam(key: String): Boolean {
        return params!!.optBoolean(key, false)
    }

    fun jsonParam(key: String): JSONObject {
        val res = strParam(key)
        return JSONObject(res)
    }

    fun lngParam(key: String): Long {
        return params!!.optLong(key)
    }

    fun callback(result: String) {
        if (null == id) {
            return
        }
        runJs(String.format("drmer.dequeue('%s', '%s');", id, result))
    }

    fun callback(model: Any?) {
        val json = when (model) {
            is JSONObject -> model.toString()
            is JSONArray -> model.toString()
            else -> Gson().toJson(model)
        }
        runJs(String.format("drmer.dequeue('%s', %s);", id, json))
    }

    private fun runJs(js: String) {
        val webView = activity.webView
        webView.post {
            if (Build.VERSION.SDK_INT >= 19) {
                webView.evaluateJavascript(js, null)
            } else {
                webView.loadUrl("javascript:$js")
            }
        }
    }

}
